relayer/strategies.go

```diff
@@ -1,58 +1,241 @@
+func QueryChannelsOnConnection(src *Chain) ([]*types.IdentifiedChannel, error) {
+	// Query the latest heights on src & dst
+	srch, err := src.ChainProvider.QueryLatestHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	// Query the list of channels for the connection on src
+	var srcChannels []*types.IdentifiedChannel
+
+	if err = retry.Do(func() error {
+		srcChannels, err = src.ChainProvider.QueryConnectionChannels(srch, src.ConnectionID())
+		return err
+	}, RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
+		src.LogRetryQueryConnectionChannels(n, err, src.ConnectionID())
+	})); err != nil {
+		return nil, err
+	}
+
+	return srcChannels, nil
+}
+
+// FilterOpenChannels takes a slice of channels and adds all the channels with OPEN state to a new slice of channels.
+// NOTE: channels will not be added to the slice of open channels more than once.
+func FilterOpenChannels(channels []*types.IdentifiedChannel, openChannels []*ActiveChannel) []*ActiveChannel {
+	inSlice := false
+
+	// Filter for open channels
+	for _, channel := range channels {
+		if channel.State == types.OPEN {
+
+			// Check if we have already added this channel to the slice of open channels
+			for _, openChannel := range openChannels {
+				if channel.ChannelId == openChannel.channel.ChannelId {
+					inSlice = true
+					break
+				}
+			}
+
+			// We don't want to add channels to the slice of open channels that have already been added
+			if !inSlice {
+				openChannels = append(openChannels, &ActiveChannel{
+					channel: channel,
+					active:  false,
+				})
+			}
+
+			inSlice = false
+		}
+	}
+
+	return openChannels
+}
+
+// RelayUnrelayedPacketsAndAcks will relay all the pending packets and acknowledgements on both the src and dst chains.
+func RelayUnrelayedPacketsAndAcks(src, dst *Chain, maxTxSize, maxMsgLength uint64, srcChannel *ActiveChannel, channels chan<- *ActiveChannel, doneChan chan struct{}) {
+	// make goroutine signal its death, whether it's a panic or a return
+	defer func() {
+		channels <- srcChannel
+	}()
+
+	for {
+		select {
+		case <-doneChan:
+			doneChan <- struct{}{}
+			return
+		default:
+			if err := RelayUnrelayedPackets(src, dst, maxTxSize, maxMsgLength, srcChannel.channel); err != nil {
+				return
+			}
+			if err := RelayUnrelayedAcks(src, dst, maxTxSize, maxMsgLength, srcChannel.channel); err != nil {
+				return
+			}
+
+			time.Sleep(1000 * time.Millisecond)
+		}
+	}
+}
+
+// RelayUnrelayedPackets fetches unrelayed packet sequence numbers and attempts to relay the associated packets.
+func RelayUnrelayedPackets(src, dst *Chain, maxTxSize, maxMsgLength uint64, srcChannel *types.IdentifiedChannel) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 45*time.Second)
+	defer cancel()
+
+	// Fetch any unrelayed sequences depending on the channel order
+	sp, err := UnrelayedSequences(src, dst, srcChannel)
+	if err != nil {
+		src.Log(fmt.Sprintf("unrelayed sequences error: %s", err))
+		return err
+	} else {
```

Comment:
The previous block of `if` returns, so drop the `else` block in order to have less indentation on the remaining code in this function.